lexxpluss_apps/src/board_controller.cpp

```diff
@@ -82,62 +82,92 @@ class power_switch_handler {  // No pins declared
+    raw_switch(const raw_switch&) = delete;
+    raw_switch & operator=(const raw_switch&) = delete;
+
     void poll() {
-        gpio_dt_spec gpio_dev = GET_GPIO(ps_sw_in);
         if (!gpio_is_ready_dt(&gpio_dev)) {
-            LOG_ERR("gpio_is_ready_dt Failed\n");
+            LOG_ERR("gpio_is_ready_dt Failed");
             return;
         }
-        int now{gpio_pin_get_dt(&gpio_dev)};
-        if (prev_raw != now) {
-            prev_raw = now;
+
+        int const now{gpio_pin_get_dt(&gpio_dev)};
+        bool const changed{prev != now};
+        prev = now;
+
+        count++;
+        if (changed) {
             count = 0;
-        } else {
-            ++count;
         }
-        bool asserted{false};
-        if (count > COUNT) {
-            count = COUNT;
-            asserted = true;
+    }
+
+    STATE get_state() {
+        if(!is_asserted()) {
+            return STATE::UNKOWN;
         }
-        if (asserted) {
-            bool changed{prev != now};
-            sw_bat.poll(changed);
-            sw_unlock.poll(changed);
-            if (changed) {
-                prev = now;
-                start_time = k_uptime_get();
-            } else if (now == 0) {
-                auto elapsed{k_uptime_get() - start_time};
-                if (elapsed > PWS_LONG_PUSHED_MS) {
-                    if (state != STATE::LONG_PUSHED)
-                        state = STATE::LONG_PUSHED;
-                } else if (elapsed > PWS_PUSHED_MS) {
-                    if (state == STATE::RELEASED)
-                        state = STATE::PUSHED;
-                }
-            }
+        
+        return prev == 0 ? STATE::LOW : STATE::HIGH;
+    }
+
+private:
+    bool is_asserted() const {
+        return  COUNT < count;
+    }
+
+    gpio_dt_spec gpio_dev;
+    uint32_t count{0};
+    int prev{-1};
+    static constexpr uint32_t COUNT{1};
+};
+
+
+#define PWS_LONG_PUSHED_MS 60000
+#define PWS_PUSHED_MS 3000
+class power_switch { // Variables Implemented
+public:
+    enum class STATE {
+        RELEASED, PUSHED, LONG_PUSHED,
+    };
+
+    void poll() {
+        raw_sw.poll();
+
+        auto const now{raw_sw.get_state()};
+        bool const changed{now != prev};
+        prev = now;
+
+        if(now == raw_switch::STATE::UNKOWN) {
+            return;
+        }
+
+        sw_bat.poll(changed);
+        sw_unlock.poll(changed);
+        if (changed) {
+            start_time = k_uptime_get();
         }
     }
     void reset_state() {
         start_time = k_uptime_get();
-        state = STATE::RELEASED;
     }
-    STATE get_state() const {return state;}
-    bool get_raw_state() {
-        gpio_dt_spec gpio_dev = GET_GPIO(ps_sw_in);
-        if (!gpio_is_ready_dt(&gpio_dev)) {
-            LOG_ERR("gpio_is_ready_dt Failed\n");
-            return false;
+    STATE get_state() const {
+        if(prev != raw_switch::STATE::LOW) {
+            return STATE::RELEASED;
```

Comment:
UNKNWONの時もRELEASED状態。

LOWからHIGHの時の応答
LOW -> STATE_PUSHED ( example )
UNKNOWN -> STATE_RELEASED
HIGH -> STATE_RELEASED

想定よりraw_switchの切り替わり分早いことになると思います。
現在stateを持っておいて、ここでは現在stateを応答する方が良いなどの方が良い可能性があります。

RELEASED state even when UNKNWON.

Response from LOW to HIGH
LOW -> STATE_PUSHED ( example )
UNKNOWN -> STATE_RELEASED
HIGH -> STATE_RELEASED

I think that the raw_switch will be switched earlier than expected.
It may be better to keep the current state and respond with the current state here.

Reply:
確認ありがとうございます。
stateの変化時はUNKNOWNを出力しないようraw_switchを修正しました。

Thanks for reviewing!!
I fixed raw_switch not to return UNKNOWN when its state is changed.

Reply:
Thank you!